File: python-algorithm/leetcode/problem_761.py
"""761. Special Binary String
https://leetcode.com/problems/special-binary-string/
"""


class Solution:
    def make_largest_special(self, s: str) -> str:
        n = len(s)
        if n <= 2:
            return s
        subs = []
        cnt = 0
        left = 0
        for i in range(n):
            if s[i] == '1':
                cnt += 1
                continue
            cnt -= 1
            if cnt == 0:
                subs.append('1' + self.make_largest_special(s[left + 1:i]) + '0')
                left = i + 1
        subs.sort(reverse=True)
        return ''.join(subs)

File: java-algo/src/main/java/com/leetcode/Problem761.java
package com.leetcode;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * 761. Special Binary String
 * https://leetcode.cn/problems/special-binary-string
 */
class Problem761 {
    public String makeLargestSpecial(String s) {
        int n = s.length();
        if (n <= 2) return s;
        int cnt = 0, left = 0;
        List<String> subs = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '1') {
                cnt++;
                continue;
            }
            if (--cnt == 0) {
                subs.add("1" + makeLargestSpecial(s.substring(left + 1, i)) + "0");
                left = i + 1;
            }
        }
        subs.sort(Comparator.reverseOrder());
        return String.join("", subs);
    }
}
